File: codes/python/ch18/eigen_benchmark_oscillator.py
#!/usr/bin/env python3
"""
eigen_benchmark_oscillator.py

Chapter 18: Linear Spectral Eigenproblems
Computational Etude 18.4: the infinite-interval tax.

Solves the quantum harmonic oscillator
        u_xx + (lambda - x^2) u = 0,   |u| -> 0 as |x| -> inf,
using the rational_chebyshev utility (algebraic map x = ell t / sqrt(1-t^2))
at two resolutions (N = 16 and N = 32) and for three map parameters.

Exact eigenvalues are lambda_j = 2 j + 1 for j = 0, 1, 2, ... with
Hermite-function eigenfunctions H_j(x) exp(-x^2/2).

Reproduces Boyd (2000) Figs 7.4 and 7.6. The pedagogical contrast with
Etude 18.3 is deliberate: for the bounded benchmark, N = 16 gives 6
trusted eigenvalues; for the unbounded benchmark, the same N buys fewer
good modes --- the "infinite-interval tax".

We also scan the map parameter ell in {2, 4, 8} to show that map-parameter
choice is a first-class tuning knob: too small an ell crowds the grid near
x = 0, too large an ell wastes nodes where the wavefunction has already
decayed.

The variable name `ell` (the Greek script letter) replaces what earlier
drafts of this textbook called `L`, to avoid collision with the truncation
half-width `L` used in chapter 20.

Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
Part of "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""
from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

SCRIPT_DIR = Path(__file__).resolve().parent
sys.path.insert(0, str(SCRIPT_DIR))
from rational_chebyshev import rational_chebyshev_derivative_matrices  # noqa: E402

plt.rcParams.update({
    "font.family": "serif",
    "font.serif": ["CMU Serif", "DejaVu Serif"],
    "mathtext.fontset": "cm",
    "font.size": 11,
    "axes.linewidth": 0.8,
    "xtick.major.width": 0.8,
    "ytick.major.width": 0.8,
    "figure.dpi": 150,
    "savefig.dpi": 300,
})

NAVY, SKY, CORAL = "#142D6E", "#7896D2", "#E74C3C"
TEAL, PURPLE, ORANGE = "#16A085", "#8E44AD", "#E67E22"

OUTPUT_DIR = SCRIPT_DIR.parent.parent.parent / "textbook" / "figures" / "ch18" / "python"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)


def solve_oscillator(N: int, ell: float):
    """Return sorted positive-real spectrum of H = -d^2/dx^2 + x^2."""
    _, D2_x, x = rational_chebyshev_derivative_matrices(N, ell)
    H = -D2_x + np.diag(x ** 2)
    lam = np.linalg.eigvals(H)
    lam = lam.real[np.abs(lam.imag) < 1e-6]
    lam = lam[lam > 0]
    return np.sort(lam)


def count_good(err: np.ndarray, tol: float = 1e-2) -> int:
    failures = np.where(err >= tol)[0]
    return failures[0] if failures.size > 0 else err.size


def make_figure(Ns=(16, 32), ell_scan=(2.0, 4.0, 8.0), ell_best: float = 4.0):
    # --- Panel A: ell = ell_best, two resolutions, error vs mode ---
    fig, axes = plt.subplots(1, 2, figsize=(11.0, 4.5))

    markers = ["o", "s"]
    colors = [NAVY, CORAL]
    ax = axes[0]
    counts_by_N = {}
    for N, m, c in zip(Ns, markers, colors):
        lam = solve_oscillator(N, ell_best)
        j = np.arange(lam.size)
        lam_exact = 2 * j + 1
        err = np.abs(lam - lam_exact)
        err_plot = np.maximum(err, 1e-17)
        ax.semilogy(j + 1, err_plot, marker=m, linestyle="-",
                    color=c, markerfacecolor="white",
                    markersize=5, linewidth=0.6, markeredgewidth=1.0,
                    label=rf"$N = {N}$, $\ell = {ell_best}$")
        counts_by_N[int(N)] = int(count_good(err, 1e-2))
    ax.axhline(1e-2, color="k", linestyle="--", lw=0.6, alpha=0.5)
    ax.set_xlabel("mode number $j$")
    ax.set_ylabel(r"$|\lambda^{\mathrm{num}}_j - \lambda^{\mathrm{exact}}_j|$")
    ax.set_title("infinite-interval tax: oscillator spectrum", fontsize=10)
    ax.set_ylim(1e-17, 1e4)
    ax.grid(True, which="both", alpha=0.25, linewidth=0.4)
    ax.legend(loc="lower right", fontsize=9)

    # --- Panel B: ell scan at fixed N = 32 ---
    ax = axes[1]
    colors_B = [SKY, NAVY, PURPLE]
    markers_B = ["^", "o", "v"]
    counts_by_ell = {}
    N_scan = 32
    for ell, m, c in zip(ell_scan, markers_B, colors_B):
        lam = solve_oscillator(N_scan, ell)
        j = np.arange(lam.size)
        lam_exact = 2 * j + 1
        err = np.abs(lam - lam_exact)
        err_plot = np.maximum(err, 1e-17)
        ax.semilogy(j + 1, err_plot, marker=m, linestyle="-",
                    color=c, markerfacecolor="white",
                    markersize=5, linewidth=0.6, markeredgewidth=1.0,
                    label=rf"$\ell = {ell}$")
        counts_by_ell[float(ell)] = int(count_good(err, 1e-2))
    ax.axhline(1e-2, color="k", linestyle="--", lw=0.6, alpha=0.5)
    ax.set_xlabel("mode number $j$")
    ax.set_ylabel(r"$|\lambda^{\mathrm{num}}_j - \lambda^{\mathrm{exact}}_j|$")
    ax.set_title(rf"$\ell$-scan at $N = {N_scan}$: map-parameter sensitivity", fontsize=10)
    ax.set_ylim(1e-17, 1e4)
    ax.grid(True, which="both", alpha=0.25, linewidth=0.4)
    ax.legend(loc="lower right", fontsize=9)

    fig.tight_layout()
    fig.savefig(OUTPUT_DIR / "eigen_benchmark_oscillator.pdf", bbox_inches="tight", pad_inches=0.05)
    fig.savefig(OUTPUT_DIR / "eigen_benchmark_oscillator.png", bbox_inches="tight", pad_inches=0.05)
    plt.close(fig)

    spectra = {
        f"N{N}_ell{ell_best}": solve_oscillator(N, ell_best).tolist() for N in Ns
    }
    spectra.update({
        f"N{N_scan}_ell{ell}": solve_oscillator(N_scan, ell).tolist() for ell in ell_scan
    })
    return {
        "Ns":              list(Ns),
        "ell_best":        ell_best,
        "ell_scan":        list(ell_scan),
        "counts_by_N":     counts_by_N,
        "counts_by_ell":   counts_by_ell,
        "spectra":         spectra,
    }


def main():
    p = argparse.ArgumentParser(description=__doc__.splitlines()[0])
    p.add_argument("--dump", type=str, default=None)
    args = p.parse_args()

    r = make_figure()
    print("[Etude 18.4]  harmonic oscillator on (-inf, +inf)")
    print(f"  ell = {r['ell_best']} scan over N:")
    for N in r["Ns"]:
        print(f"    N = {N:2d}:  {r['counts_by_N'][N]} good eigenvalues (|err| < 1e-2)")
    print(f"  N = 32 scan over ell:")
    for ell in r["ell_scan"]:
        print(f"    ell = {ell}:  {r['counts_by_ell'][ell]} good eigenvalues")
    print(f"  figure: {OUTPUT_DIR / 'eigen_benchmark_oscillator.pdf'}")
    if args.dump:
        Path(args.dump).write_text(json.dumps(r, indent=2))
        print(f"  dumped: {args.dump}")


if __name__ == "__main__":
    main()

File: codes/julia/ch18/eigen_benchmark_oscillator.jl
# eigen_benchmark_oscillator.jl
#
# Chapter 18: Linear Spectral Eigenproblems
# Computational Etude 18.4: the infinite-interval tax.
#
# Solves u_xx + (lambda - x^2) u = 0, u -> 0 as |x| -> inf, using the
# rational Chebyshev algebraic map. Exact spectrum lambda_j = 2j+1,
# eigenfunctions H_j(x) exp(-x^2/2).
#
# Reproduces Boyd (2000) Figs 7.4 and 7.6.
#
# Author: Dr. Denys Dutykh (Khalifa University of Science and Technology)
# Part of "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

include(joinpath(@__DIR__, "rational_chebyshev.jl"))
using .RationalChebyshev: rational_chebyshev_derivative_matrices

_json_val(x::Bool)          = x ? "true" : "false"
_json_val(x::Integer)       = string(x)
_json_val(x::Real)          = isfinite(x) ? string(Float64(x)) : (isnan(x) ? "NaN" : (x > 0 ? "Infinity" : "-Infinity"))
_json_val(x::AbstractString)= "\"" * replace(String(x), "\\" => "\\\\", "\"" => "\\\"") * "\""
_json_val(v::AbstractVector)= "[" * join(_json_val.(v), ", ") * "]"
function _json_val(d::AbstractDict)
    pairs = String[]
    for (k, v) in d; push!(pairs, _json_val(string(k)) * ": " * _json_val(v)); end
    return "{" * join(pairs, ", ") * "}"
end
write_json(path, x) = open(io -> print(io, _json_val(x)), path, "w")

set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 11,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 9,),
))

const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const PURPLE = colorant"#8E44AD"

const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch18", "julia")
mkpath(OUTPUT_DIR)

function solve_oscillator(N::Int, L::Real)
    _, D2_x, x = rational_chebyshev_derivative_matrices(N, L)
    H = -D2_x + Diagonal(x .^ 2)
    lam = eigvals(H)
    keep = abs.(imag.(lam)) .< 1e-6
    lam = real.(lam[keep])
    lam = lam[lam .> 0]
    return sort(lam)
end

function count_good(err, tol = 1e-2)
    for (k, e) in enumerate(err)
        e >= tol && return k - 1
    end
    return length(err)
end

function make_figure(; Ns = (16, 32), L_scan = (2.0, 4.0, 8.0), L_best::Real = 4.0,
                     dump_path = nothing)
    fig = Figure(size = (1100, 450))

    # Panel A: N-scan at fixed L_best
    ax1 = Axis(fig[1, 1];
        xlabel = "mode number j",
        ylabel = "|λ_num − λ_exact|",
        title  = "infinite-interval tax: oscillator spectrum",
        yscale = log10, limits = (nothing, (1e-17, 1e4)))
    markers = [:circle, :rect]
    cols = [NAVY, CORAL]
    counts_by_N = Dict{Int,Int}()
    for (N, m, c) in zip(Ns, markers, cols)
        lam = solve_oscillator(N, L_best)
        j = collect(0:length(lam)-1)
        lam_exact = 2 .* j .+ 1
        err = abs.(lam .- lam_exact)
        err_plot = max.(err, 1e-17)
        scatterlines!(ax1, j .+ 1, err_plot;
                      color = c, markercolor = :white,
                      strokecolor = c, strokewidth = 1.0,
                      linewidth = 0.6, marker = m, markersize = 8,
                      label = "N = $N, L = $L_best")
        counts_by_N[N] = count_good(err, 1e-2)
    end
    hlines!(ax1, [1e-2]; color = :black, linestyle = :dash,
            linewidth = 0.6, alpha = 0.5)
    axislegend(ax1, position = :rb, labelsize = 9, patchsize = (12, 12))

    # Panel B: L-scan at fixed N
    ax2 = Axis(fig[1, 2];
        xlabel = "mode number j",
        ylabel = "|λ_num − λ_exact|",
        title  = "L-scan at N = 32: map-parameter sensitivity",
        yscale = log10, limits = (nothing, (1e-17, 1e4)))
    markers2 = [:utriangle, :circle, :dtriangle]
    cols2 = [SKY, NAVY, PURPLE]
    counts_by_L = Dict{Float64,Int}()
    N_scan = 32
    for (L, m, c) in zip(L_scan, markers2, cols2)
        lam = solve_oscillator(N_scan, L)
        j = collect(0:length(lam)-1)
        lam_exact = 2 .* j .+ 1
        err = abs.(lam .- lam_exact)
        err_plot = max.(err, 1e-17)
        scatterlines!(ax2, j .+ 1, err_plot;
                      color = c, markercolor = :white,
                      strokecolor = c, strokewidth = 1.0,
                      linewidth = 0.6, marker = m, markersize = 8,
                      label = "L = $L")
        counts_by_L[L] = count_good(err, 1e-2)
    end
    hlines!(ax2, [1e-2]; color = :black, linestyle = :dash,
            linewidth = 0.6, alpha = 0.5)
    axislegend(ax2, position = :rb, labelsize = 9, patchsize = (12, 12))

    save(joinpath(OUTPUT_DIR, "eigen_benchmark_oscillator.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "eigen_benchmark_oscillator.png"), fig, px_per_unit = 4)

    spectra = Dict{String,Vector{Float64}}()
    for N in Ns
        spectra["N$(N)_L$(L_best)"] = solve_oscillator(N, L_best)
    end
    for L in L_scan
        spectra["N$(N_scan)_L$(L)"] = solve_oscillator(N_scan, L)
    end

    r = Dict{String,Any}(
        "Ns"            => collect(Ns),
        "L_best"        => L_best,
        "L_scan"        => collect(L_scan),
        "counts_by_N"   => Dict(string(k) => v for (k, v) in counts_by_N),
        "counts_by_L"   => Dict(string(k) => v for (k, v) in counts_by_L),
        "spectra"       => spectra,
    )
    dump_path !== nothing && write_json(dump_path, r)
    return r, counts_by_N, counts_by_L
end

function parse_args()
    dump_path = nothing; i = 1
    while i <= length(ARGS)
        if ARGS[i] == "--dump" && i < length(ARGS); dump_path = ARGS[i+1]; i += 2
        else i += 1 end
    end
    return dump_path
end

if abspath(PROGRAM_FILE) == @__FILE__
    dp = parse_args()
    _, cN, cL = make_figure(; dump_path = dp)
    println("[Etude 18.4]  harmonic oscillator on (-inf, +inf)")
    println("  L = 4.0 scan over N:")
    for N in sort(collect(keys(cN)))
        @printf("    N = %2d:  %d good eigenvalues (|err| < 1e-2)\n", N, cN[N])
    end
    println("  N = 32 scan over L:")
    for L in sort(collect(keys(cL)))
        @printf("    L = %.1f:  %d good eigenvalues\n", L, cL[L])
    end
    println("  figure: ", joinpath(OUTPUT_DIR, "eigen_benchmark_oscillator.pdf"))
    dp !== nothing && println("  dumped: ", dp)
end
